ui: Bugfix several bug fixes around cluster version create volume

ChengYanJin opened the pull request:
**Component**: ui

**Context**: 
Bugfix:
1. When the size is `Unknown` no need to add a unit
2. Add the color of `create` button in create node & volume page to `buttonPrimary`
3. Fix the cluster version in create node page
4. Add about entry from the Navbar
5. Fix the display bug of labels in crate volume page
6. Add default `ui_base_path` in `config.json`
7.  Translations 
8. After deleting a volume, redirect to the first volume - overview tab
9. Fix the error when a new node is not ready, we can receive the string from ipInterfaceObject

Improvement:
10. Merge the label validation before creating a new volume and **improve the UX of creating volume**
- when the label key and value are valid, the `Add` button will become enabled, otherwise stay disable.

![image](https://user-images.githubusercontent.com/18453133/117105374-2dc06480-ad7e-11eb-9561-d5f3d85ce898.png)



**Summary**:

**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3303  
Closes: #3328
Closes: #3203
<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Commit:
ui: Update volume label validation for kubernetes

When creating a volume using specials characters in the UI is authorised for
labels but denied when submitting the query to kubernetes API.
In the current version, label is validated and display a message if the user
fills label with invalid characters

Refs: #3203

ChengYanJin committed:
ui: Add regex check before add unit in formatSizeForDisplay

To fix the bug that we add `B` for unknown size as well

ChengYanJin committed:
ui: Change the create button to buttonPrimary

ChengYanJin committed:
ui: Fix the cluster version in create node and about page

We couldn't retrieve the cluster version is because we call the
fetchClusterAction before we are authenticated to k8s api.

Solution: since we use the `cluster version` only in create node page
and about page, it makes sense to fetch the data only in those two
pages.

ChengYanJin committed:
salt: Add about entry in navbar-subLogin

ChengYanJin committed:
ui: Fix the bug of around create volume labels field

UX improvement:
When the labelName and labelValue are not empty, we change the color of
the button to indicate clicking on the `Add` button is necessary.

Refs: #3303 #3328

ChengYanJin committed:
ui: Add ui_base_path in the config.json

To fix the broken history when running the UI locally

ChengYanJin committed:
ui/translation: Add the plural volume

ChengYanJin committed:
ui: Display the overview tab after deleting a volume

ChengYanJin committed:
ui: ipsInterfacesObject maybe be a string with error message

This bug breaks the entire reducer :(
Again, we should fetch this ips info through k8s API instead of Salt

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -568,14 +570,25 @@ const CreateVolume = (props) => {
+                              variant={'buttonSecondary'}
+                            />
+                          ) : (
+                            <Button
+                              text={intl.translate('add')}
+                              type="button"
+                              onClick={addLabel}
+                              data-cy="add-volume-labels-button"
+                              outlined
+                            />
```
shouldn't we disable the button initially instead of just changing it's color ? Also I think we can remove the onClick handler in this case

ChengYanJin replied:
Done in https://github.com/scality/metalk8s/pull/3338/commits/7a961e20d30abe2e3ea76d4d7050fbe7bc9464c4 ;)

JBWatenbergScality commented on ui/src/services/salt/api.js:
```diff
@@ -68,10 +68,36 @@ export type IPInterfaces = {
+          ]
+        }
+      },
+      "toto": "Minion did not return. [No response]\nThe minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:\n\nsalt-run jobs.lookup_jid 20210429184803777520"
```
```suggestion
      "nodeName": "Minion did not return. [No response]\nThe minions may not have all finished running and any remaining minions will return upon completion. To look up the return data for this job later, run the following command:\n\nsalt-run jobs.lookup_jid 20210429184803777520"
```

ChengYanJin committed:
ui: Fix the title of active alerts

ChengYanJin committed:
ui: Set the button disable when it's not ready

Refs: #3303

ChengYanJin committed:
salt: Fix locale in salt master container

ChengYanJin committed:
Merge pull request #3278 from scality/improvement/3203-volume-label-validation

ui: Update volume label validation for kubernetes

ChengYanJin committed:
ui: Add label validation to the Yup schema

TODO:
We display only one error message for labelName. However, when there is
prefix in the labelName the error message should be more specific.

Call handleErrorLabel in validate prop in labelName <Field/>

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -435,13 +453,64 @@ const CreateVolume = (props) => {
                   }
                 }
               };
+              // TODO: Apply this function as `labelName` field validation function, for the moment we only have one error message.
```
Is this todo still valid ? I see various error message in the implementation.

ChengYanJin replied:
OK. So in the end I switched to `Field-level Validation` instead of `Yup` schema which will display the error message separately for the `label prefix` and `labe name value`.
Please see: https://github.com/scality/metalk8s/pull/3338/commits/9917a17a96db87f73451d6d13ff38e6b7db8de77

ChengYanJin committed:
ui: Use validate function to validate labelName instead of Yup schema

ChengYanJin committed:
ui: Add interaction test for create volume form

JBWatenbergScality commented on ui/src/containers/CreateVolume.js:
```diff
@@ -453,64 +455,40 @@ const CreateVolume = (props) => {
-                  values.labelValue,
-                );
 
+              const validataLabelName = (value) => {
```
validateLabelName

JBWatenbergScality commented on ui/cypress/fixtures/kubernetes/storageclasses.json:
```diff
@@ -0,0 +1,207 @@
+{
+  "apiVersion": "storage.k8s.io/v1",
+  "items": [
```
I would limit the storage classes only to the one that makes sense for our tests

ChengYanJin replied:
OK. I was hesitating should I keep only one storage class. I will do so then.

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,70 @@
+  it('should be enabled and the Add button should be disabled when no labels is specified', () => {
+    cy.get('input[name=name]').type(VOLUME_NAME);
+
+    cy.get('.sc-select').eq(0).click();
+    cy.get('.sc-select__menu').eq(0).click();
+    cy.get('.sc-select').eq(1).click();
+    cy.get('.sc-select__menu')
```
We should avoid dom selectors by classnames and tests this as user interact with. With the current implementation the most elegant solution is to look for elements with text matching /node*/i and click on the next `.sc-select`. Ideally we should add an aria role with listbox value on the dropdown component https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role + an aria-label with value "node"

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,70 @@
+    cy.get('.sc-select__menu').eq(0).click();
+    cy.get('.sc-select').eq(1).click();
+    cy.get('.sc-select__menu')
+      .find(`[data-cy=storageClass-${STORAGECLASS}]`)
```
We should select the element via the text presented to the user

ChengYanJin replied:
In the end, we got to keep the `data-cy`, the reason is:

>   /* we use `data-cy`is because the storageClass may display in both the select list and selected item, 
>      so we use unique selector to make sure to choose the item in the list. */

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,70 @@
+    cy.get('.sc-select__menu').find(`[data-cy="type-${VOLUME_TYPE}"]`).click();
+    cy.get('input[name=sizeInput]').type(VOLUME_SIZE);
+
+    cy.get('[data-cy="submit-create-volume"]').should('be.enabled');
```
use cy.queryByRole('button', {name: /create/i})

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,70 @@
+    cy.get('input[name=sizeInput]').type(VOLUME_SIZE);
+
+    cy.get('[data-cy="submit-create-volume"]').should('be.enabled');
+    cy.get('[data-cy="add-volume-labels-button"]').should('be.disabled');
```
use cy.queryByRole('button', {name: /add/i})

ChengYanJin replied:
Just FYI in case you miss it :)
https://github.com/testing-library/cypress-testing-library/releases/tag/v6.0.0
so should be: `  cy.findByRole('button', { name: /create/i })`

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/several-bug-fixes-cluster-version-create-volume-labels` into
`development/2.9`
* [w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels](https://github.com/scality/metalk8s/compare/development/2.10...w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/several-bug-fixes-cluster-version-create-volume-labels` into
`development/2.9`
* [w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels](https://github.com/scality/metalk8s/compare/development/2.10...w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @jbwatenbergscality

* @teddyandrieux

* @gdemonet

* @alexandre-allard-scality

* @thomasdanan

* @lucieleonard

bert-e commented:
# History mismatch



Merge commit #47d134654a74cd457e0dafb10e73ed71bd52f158 on the integration branch
`w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels` is merging a branch which is neither the current
branch `bugfix/several-bug-fixes-cluster-version-create-volume-labels` nor the development branch
`development/2.10`.

It is likely due to a rebase of the branch `bugfix/several-bug-fixes-cluster-version-create-volume-labels` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,59 @@
+  });
+
+  it('should be disabled before adding the valid label', () => {
+    cy.get('input[name=labelName]').type(VOLUME_LABEL_NAME);
```
it could be replaced with findByRole('textbox', {name: /Enter label name/i})

ChengYanJin replied:
OK. Maybe` cy.findByPlaceholderText(/Enter label name/i).` would be even better!

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,59 @@
+
+  it('should be disabled before adding the valid label', () => {
+    cy.get('input[name=labelName]').type(VOLUME_LABEL_NAME);
+    cy.get('input[name=labelValue]').type(VOLUME_LABEL_VALUE);
```
it could be replaced with findByRole('textbox', {name: /Enter label value/i})

JBWatenbergScality commented on ui/cypress/integration/volume/createvolume.spec.js:
```diff
@@ -0,0 +1,59 @@
+  });
+
+  it('should be disabled as well as the Add button if labelName is not valid', () => {
+    cy.get('input[name=labelName]').type(VOLUME_LABEL_INVALID);
+    cy.get('input[name=labelValue]').type(VOLUME_LABEL_VALUE);
```
same here

ChengYanJin commented:
/status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# History mismatch



Merge commit #47d134654a74cd457e0dafb10e73ed71bd52f158 on the integration branch
`w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels` is merging a branch which is neither the current
branch `bugfix/several-bug-fixes-cluster-version-create-volume-labels` nor the development branch
`development/2.10`.

It is likely due to a rebase of the branch `bugfix/several-bug-fixes-cluster-version-create-volume-labels` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/several-bug-fixes-cluster-version-create-volume-labels` into
`development/2.9`
* [w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels](https://github.com/scality/metalk8s/compare/development/2.10...w/2.10/bugfix/several-bug-fixes-cluster-version-create-volume-labels) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @chengyanjin

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @nicolast

* @thomasdanan

* @lucieleonard

ChengYanJin commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/177ebc21e1798c7da91f4e8df927b468a19aa7d1) did not succeed in branch bugfix/several-bug-fixes-cluster-version-create-volume-labels.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.9`

* :heavy_check_mark: `development/2.10`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.9`

* :heavy_check_mark: `development/2.10`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`



Please check the status of the associated issue None.

Goodbye chengyanjin.
